lib/authentication/auth_okta.js

```diff
@@ -49,49 +51,64 @@ function AuthOkta(password, region, account, clientType, clientVersion, httpClie
   * @returns {null}
   */
   this.authenticate = async function (authenticator, serviceName, account, username) {
-    let ssoUrl;
-    let tokenUrl;
-    await step1(authenticator, serviceName, account, username).then((response) => {
-      const responseData = response['data'];
-      const success = responseData['success'];
-      const errorCode = responseData['code'];
-      const errorMessage = responseData['message'];
-
-      if (typeof success === 'undefined' || errorCode === 'undefined' || errorMessage === 'undefined') {
-        throw new Error('Unable to use provided Okta address as an authenticator. Is the authenticator URL correct?');
-      }
+    const response = await getAuthURLs(authenticator, serviceName, account, username);
+    const responseData = response['data'];
+    const success = responseData['success'];
+    const errorCode = responseData['code'];
+    const errorMessage = responseData['message'];
+    user = username;
 
-      if (success !== true) {
-        throw new Error(`Unable to use provided Okta address as an authenticator. Error code: ${errorCode}, error message: ${errorMessage}`);
-      }
+    if (typeof success === 'undefined' || errorCode === 'undefined' || errorMessage === 'undefined') {
+      throw new Error('Unable to use provided Okta address as an authenticator. Is the authenticator URL correct?');
+    }
 
-      ssoUrl = responseData['data']['ssoUrl'];
-      tokenUrl = responseData['data']['tokenUrl'];
-    });
+    if (success !== true) {
+      throw new Error(`Unable to use provided Okta address as an authenticator. Error code: ${errorCode}, error message: ${errorMessage}`);
+    }
 
-    step2(authenticator, ssoUrl, tokenUrl);
+    ssoUrl = responseData['data']['ssoUrl'];
+    tokenUrl = responseData['data']['tokenUrl'];
 
-    let oneTimeToken;
-    await step3(tokenUrl, username, password).then((response) => {
-      const data = response['data'];
+    validateURLs(authenticator, ssoUrl, tokenUrl);
+    
+    const responseHtml = await getSAMLResponse( await createAccessToken(tokenUrl, username, password), ssoUrl);
 
-      if (data['sessionToken']) {
-        oneTimeToken = data['sessionToken'];
-      } else {
-        oneTimeToken = data['cookieToken'];
-      }
-    });
+    validateSAML(responseHtml);
+  };
+
+  this.reauthenticate = async function (body, retryOption) {
+    const maxRetryTimeout = connectionConfig.getRetryTimeout();
+    const maxRetryCount = connectionConfig.getRetrySfMaxLoginRetries();
+    const remainingTimeout =  (maxRetryTimeout - retryOption.totalElapsedTime) * 1000;
 
+    const startTime = Date.now();
     let responseHtml;
-    await step4(oneTimeToken, ssoUrl).then((response) => {
-      responseHtml = response['data'];
-    });
 
-    step5(responseHtml);
-  };
+    while ((maxRetryTimeout === 0 || Date.now() < startTime + remainingTimeout) && retryOption.numRetries <= maxRetryCount) {
```

Comment:
Now, in case of maxRetryTimeout = 0 & maxRetryCount = 0 we will still perform a retry, right? Overall I think we should have a unit test that checks just this condition in isolation.

Reply:
Hello @sfc-gh-dheyman-1, I understand what you worried about. However, this is a little complicated problem.
If maxRetryTimeout = 0 & maxRetryCount = 0, the driver will not perform the retry because of the point of the retry logic because the minimum value of the numRetries is always 1.

The workflow is

1. obtain the sleep time with the jitter Util function
2. Increased the numRetries value
3. Reauthenticated in Okta.

As a result, if the maxRetryCount = 0, the retryOption.numRetries is always greater than maxRetryCount, the authentication will not be executed. Also, this authentication cannot be executed because of the condition before we start the retry logic in the SF. (https://github.com/snowflakedb/snowflake-connector-nodejs/pull/745/files#diff-333982e8b49a2955b892e81f109de2f00423a1cdbe166258fb1eded7a6c906e2L1101).

If the maxRetryCount = 1, the driver will execute 1 retry because my condition in the reauthenticate is  `retryOption.numRetries **<=** maxRetryCount`.

Therefore, this will not be a problem.

Reply:
FYI, I changed one part. In the SF, I replaced the value of numtries 1 to 0. When I implemented the new retry strategy in the past. I just used the same number. However, I found out we should start at 0, not 1 when I answered to your question.
Please have a llok

Reply:
It seems like because of the change 1 to 9, the OCSP testing is taking longer than the previous. When the numRetries is 1, it might run 6 retries although the max retry is 7. I extend the testing time for OCSP integration testing because 90 seconds are too short to test the OCSP-fail.

Reply:
I think we can leave it like it is, your explanation makes sense. By unit testing this condition, I meant extracting the condition to a method like `boolean shouldPerformAuthRetry(retryCount, remainingTimeout, etc.)` and testing this method. Apart from that the PR looks fine.

Reply:
@sfc-gh-dheyman-1, ah I see. I created a function shouldRetryOktaAuth in the util and made the unit tests. Please have a look.

Reply:
Thanks 👌